gix-transport/src/client/blocking_io/file.rs

```diff
@@ -126,6 +126,15 @@ impl client::TransportWithoutIO for SpawnProcessOnDemand {
     }
 }
 
+impl Drop for SpawnProcessOnDemand {
+    fn drop(&mut self) {
+        if let Some(mut child) = self.child.take() {
+            child.kill().ok();
```

Comment:
I'd feel better if we could remove the `kill()` then. Despite being aware that it would probably prevent to hang in the `wait()` call, I'd only want to add it when it's proven to be necessary. Since Git doesn't seem to be doing that, I think neither should we.

Reply:
I've removed it in my most recent update. I also confirmed that with this change, my application (a long-running daemon that uses gitoxide to do tons of git clones) does not show any zombied ssh processes.

Reply:
Looks like I spoke too soon! Seems there's some tests where removing the `kill()` is causing a hang. Let me look into this...

Reply:
Thanks for running this experiment!

In this case, Git might not even have all the answers as it would never encounter such a case, after all, it can't be used as a library. 

Something that surprises me though is how `wait()` (without `kill()`) can still leave zombies - if it doesn't block forever then it should shut down the child for good.

Reply:
> Something that surprises me though is how `wait()` (without `kill()`) can still leave zombies - if it doesn't block forever then it should shut down the child for good.

Not sure what you mean... The issue with the test hang is that `wait()` **is** blocking forever. After adding the `wait()`, we are guaranteed to never have zombies (at least for this child process scenario) - we're just exposed to a deadlock possibility if the `ssh` process didn't shutdown gracefully.

I was able to fix the test hang with this change:
```
diff --git a/gix/src/remote/connection/fetch/receive_pack.rs b/gix/src/remote/connection/fetch/receive_pack.rs
index 7634b34cf..aeaf17033 100644
--- a/gix/src/remote/connection/fetch/receive_pack.rs
+++ b/gix/src/remote/connection/fetch/receive_pack.rs
@@ -253,6 +253,10 @@ where
                         .transpose()?
                         .unwrap_or(false);
                     if reject_shallow_remote {
+                        drop(reader);
+                        gix_protocol::indicate_end_of_interaction(&mut con.transport, con.trace)
+                            .await
+                            .ok();
                         return Err(Error::RejectShallowRemote);
                     }
                     shallow_lock = acquire_shallow_lock(repo).map(Some)?;
```

But... now that I look at it more, I feel like there's other scenarios where tke `kill()` is going to be needed. For example, while we're in the middle of reading the pack, the user can abort the operation via `should_interrupt`. If that happens, then there is no "graceful" way to indicate to the remote that we'd like to shutdown (AFAIK).

Thoughts @Byron ?

Reply:
Right, I didn't know that deadlocks are actually happening then, and it makes sense to see these until the remote hangs up. Oh, and it looks like CI is already deadlocking.

And it's true, once interrupted, everything winds down and it's unclear in which state `ssh` is in that moment, so a kill() call would be required. Git is different, as in doesn't have that problem, as they just abort on signal. There is some special handling for tempfiles, but that's about it.

Thus, we really have to call kill here. Could you protect that `kill()` call with a comment that briefly explains why?

Then I think this can be merged, and we are back to were we were, but with a comment and a better understanding :).

Reply:
Done. Thanks for the feedback!